src/common/src/array/map_array.rs

```diff
@@ -0,0 +1,294 @@
+        self.as_scalar_ref().fmt(f)
+    }
+}
+
+impl Display for MapValue {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.as_scalar_ref().write(f)
+    }
+}
+
+impl MapValue {
+    pub fn from_list_entries(list: ListValue) -> Self {
+        if cfg!(debug_assertions) {
+            // validates list type is valid
+            _ = MapType::from_list_entries(list.data_type());
+        }
+        // TODO: validate the values is valid
+        MapValue(list)
+    }
+
+    pub fn from_kv(keys: ListValue, values: ListValue) -> Self {
+        if cfg!(debug_assertions) {
+            assert_eq!(
+                keys.len(),
+                values.len(),
+                "keys: {keys:?}, values: {values:?}"
+            );
+            let unique_keys = keys.iter().unique().collect_vec();
+            assert!(
+                unique_keys.len() == keys.len(),
+                "non unique keys in map: {keys:?}"
+            );
+            assert!(!unique_keys.contains(&None), "null key in map: {keys:?}");
+        }
+
+        let len = keys.len();
+        let key_type = keys.data_type();
+        let value_type = values.data_type();
+        let struct_array = StructArray::new(
+            MapType::struct_type_for_map(key_type, value_type),
+            vec![keys.into_array().into_ref(), values.into_array().into_ref()],
+            Bitmap::ones(len),
+        );
+        MapValue(ListValue::new(struct_array.into()))
+    }
+}
+
+/// A map is just a slice of the underlying struct array.
+///
+/// XXX: perhaps we can make it `MapRef<'a, 'b>(ListRef<'a>, ListRef<'b>);`.
+/// Then we can build a map ref from 2 list refs without copying the data.
+/// Currently it's impossible.
+#[derive(Copy, Clone, Eq)]
+pub struct MapRef<'a>(pub(crate) ListRef<'a>);
+
+impl<'a> MapRef<'a> {
+    /// Iterates over the elements of the map.
+    pub fn iter(
+        self,
+    ) -> impl DoubleEndedIterator + ExactSizeIterator<Item = (ScalarRefImpl<'a>, DatumRef<'a>)> + 'a
+    {
+        self.0.iter().map(|list_elem| {
+            let list_elem = list_elem.expect("the list element in map should not be null");
+            let struct_ = list_elem.into_struct();
+            let (k, v) = struct_
+                .iter_fields_ref()
+                .next_tuple()
+                .expect("the struct in map should have exactly 2 fields");
+            (k.expect("map key should not be null"), v)
+        })
+    }
+}
+
+impl Debug for MapRef<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_list().entries(self.0.iter()).finish()
+    }
+}
+
+impl ToText for MapRef<'_> {
+    fn write<W: std::fmt::Write>(&self, f: &mut W) -> std::fmt::Result {
+        // Note: This is arbitrarily decided...
+        write!(
+            f,
+            "{{{}}}",
+            self.iter().format_with(",", |(key, value), f| {
+                let key = key.to_text();
+                let value = value.to_text();
```

Comment:
Shall we pass the datatype of key and value?

Reply:
Do you mean `to_text_with_type`? It's actually unnecessary. lol 

https://github.com/risingwavelabs/risingwave/pull/17697/files#diff-e2c25e8efaa9e301065f987bb2e858108a22d9058ee1e094c5ef9460048291dbR59-R62

Reply:
Actually I am still thinking about keeping `to_text_with_type` and deprecate/remove `to_text`. It is just a coincident that currently these types display without full type info. Compare it with other signatures:
* `ToBinary::to_binary_with_type` (and `postgres_types::ToSql::to_sql`)
* from binary with type (not a trait yet)
* from text with type (not a trait yet)

Reply:
let's discuss in #17817